flight/mcu/lib/FifoBuffer/FifoBuffer.h

```diff
@@ -1,67 +1,163 @@
-//
-// Created by 16182 on 9/27/2021.
-//
-
 #ifndef TARS_SOFTWARE_FIFO_H
 #define TARS_SOFTWARE_FIFO_H
 
-#include <cstdint>
-#include <cstring>
+#include <ChRt.h>
 
-#ifdef COMPILE_LOCAL
-#include <mutex>
-#endif
+template<typename T, size_t max_size>
+class FifoView;
 
-#ifdef COMPILE_TARGET
-#include <ChRt.h>
-#endif
-
-// only works with types that can be copied with memcpy
-class GenericFifoBuffer {
-   public:
-    // buff_size = size of array in elements
-    // data_size = size of array element in bytes
-    GenericFifoBuffer(void* arr, uint16_t buff_size, uint16_t data_size) {
-        capacity_ = buff_size;
-        cur_length_ = 0;
-        head_idx_ = 0;
-        tail_idx_ = 0;
-        data_size_ = data_size;
-        arr_ = arr;
+template<typename T, size_t max_size>
+class FifoBuffer {
+public:
+    friend class FifoView<T, max_size>;
+
+    MUTEX_DECL(lock);
+
+    FifoBuffer() = default;
+
+    bool push(T const& element) {
+        chMtxLock(&lock);
+        arr[tail_idx++] = element;
+        if (tail_idx == max_size) {
+            tail_idx = 0;
+        }
+        count++;
+        chMtxUnlock(&lock);
+        return true;
+    }
+
+    bool read(T& read_to) {
+        chMtxLock(&lock);
+        if (count < 1) {
+            return false;
+        }
+        read_to = arr[tail_idx];
+        chMtxUnlock(&lock);
+        return true;
+    }
```

Comment:
I'm confused, read should be reading the head element, no?

Reply:
also the mutex logic is wrong

Reply:
Should be fixed